doc/adr/0018-synchronising-queries-across-indexers-running-in-parallel.rst

```diff
@@ -0,0 +1,72 @@
+
+I think this solution is better than what we currently have because it gives a principled way of composing queries over a set of indexers that run at different speeds. This allows us to run them in parallel on the same processor (as we do now) or distributed. It allows for efficient use of resources since we don't need to use any locking. The implementation is also simple since we don't share any data across threads and we don't need to deal with thead synchronisation issues.
+
+Implications
+------------
+
+However, this solution also complicates the life of our users. Now, they need to also have some function that allows them to fold these differences into the older query responses. I would suggest a couple of strategies that we could employ to simplify their lives, that are not the object of this ADR:
+
+1. Write a set of property-based tests that validate that the folding function is correct and instruct our users to verify their implementation using our acceptance tests.
+2. Develop a new software service that can synchronise queries across several indexers. This service would use our implementation of the folding function (which is verified by our tests) and after receiving all the query responses will select the best ones and provide them to the users already folded.
+
+The only thing remaining is to note that if we want to protect against rollbacks we will also need to store the block hash together with the slot number and at this point our design is complete.
+
+Alternatives
+------------
+
+Currently, the way we almost avoid that situation is to run a set of indexers in parallel by waiting on all of them to finish processing the current block before advancing to the next one. This is a fairly big improvement, but essentially it still has the same problem, except that it will not be triggered as often. One scenario when this could happen is if the query comes while the indexers are about to process a block. Some of these indexers may answer the query before block processing, and some after the block has been processed. This can be mitigated by a query-locking mechanism that spans all the indexers, but this adds complexity while not managing to solve the root problem.
+
+This solution also forgoes the possibility of splitting up the indexers into different processes or machines. Our customers complained about the excessive resource consumption, performance and the inability to easily add their own indexers to the chain index. Giving users the possibility of distributing their indexers will partially address the first concern by splitting resource consumption on several machines (which is also addressed at the indexer design level by an inherently scalable design) and indirectly the second (since indexers are distributable, they have a simple, standard API that allows any customer to roll their own version, implicitly solving the hard problems like rollbacks and synchronisation).
```

Comment:
forgoes => allows?

AFAIK the thing that takes time is building up the index, and after that the indexers just tick along one block at a time while consuming little resources.

I would also contest the idea that rollbacks are a hard problem (to us), as to me it seems you just roll back (= delete newer rows from the sqlite) and that is the end of it, this is what we do, no?

It seems to me that it is our clients instead who are the ones that *will continue to have the hard problem* of choosing how bleeding edge they want to be (= how often they will get rollbacks, the probability declines over time) and how will they handle them. But the choice is theirs, no?

E.g, for being totally bleeding edge I guess the solution is some kind of layer 2 like maybe roll-ups. Say if Djed wants to do some atomic thing, then even though that thing might look like multiple things to them, they still collect it all into a single transaction and submit that. That single transaction will either succeed or be rolled back, but in both cases the final state is consistent (this being not much different in how relational databases have transactions).

I wonder how bleeding edge either Mamba or Djed wants to be, and if they are totally bleeding edge then what do they do when they get a rollback? @koslambrou

Reply:
For now, the Sidechain team doesn't need bleeding edge. They actually only the stable blocks, so they don't care for rollbacks. For Djed, they are bleeding edge, but don't care for rollbacks because their contracts are one-shot (the transaction will just fail in the user's interface)